File: easy/Closest-Binary-Search-Tree-Value_easy/Solution1.java
/*
Given a non-empty binary search tree and a target value, find the value in the BST that is closest to the target.

Note:
Given target value is a floating point.
You are guaranteed to have only one unique value in the BST that is closest to the target.
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

    public int closestValue(TreeNode root, double target) {
        int res = root.val;
        Stack<TreeNode> stack = new Stack<TreeNode>();
        while(!stack.isEmpty() || root != null) {
            if(root != null) {
                stack.push(root);
                root = root.left;
            } else {
                root = stack.pop();
                if(target < (double)root.val) {
                    if(Math.abs(target - (double)res) >= Math.abs(target - (double)root.val)) {
                        res = root.val;
                    } else {
                        return res;
                    }
                } else if (target > root.val) {
                    res = root.val;
                } else {
                    return root.val;
                }
                root = root.right;
             }
        }
        return res;
    }
}

File: algorithms/Longest_Mountain_in_Array-medium/solution1.cpp
class Solution {
public:
    int longestMountain(vector<int>& arr) {
        int n = arr.size();
        int res = 0, left = 0;
        while (left < n) {
            int right = left;
            if (right + 1 < n && arr[right] < arr[right+1]) {
                while (right + 1 < n && arr[right] < arr[right+1]) right++;
                if (right + 1 < n && arr[right] > arr[right+1]) {
                    while (right + 1 < n && arr[right] > arr[right+1]) right++;
                    res = max(res, right - left + 1);
                }
            }
            left = max(right, left + 1);
        }
        return res;
    }
};
